File: src/main/scala/is/hail/stats/Distributions.scala
package is.hail.stats

import breeze.stats.distributions._
import is.hail.utils._

sealed trait Distribution {
    def getBreezeDist(randBasis: RandBasis): Rand[Double]
}
case class UniformDist(min: Double, max: Double) extends Distribution {
  if (min >= max)
    fatal(s"min $min must be less than max $max")

  override def getBreezeDist(randBasis: RandBasis): Rand[Double] = {
    new Uniform(min, max)(randBasis)
  }
}

case class BetaDist(a: Double, b: Double) extends Distribution {
  if (a <= 0 || b <= 0)
    fatal(s"a and b must be positive, got $a and $b")

  override def getBreezeDist(randBasis: RandBasis): Rand[Double] = {
    new Beta(a, b)(randBasis)
  }
}

case class TruncatedBetaDist(a: Double, b: Double, min: Double, max: Double) extends Distribution {
  if (min >= max)
    fatal(s"min $min must be less than max $max")

  override def getBreezeDist(randBasis: RandBasis): Rand[Double] = {
    new TruncatedBeta(a, b, min, max)(randBasis)
  }
}


class TruncatedBeta(a: Double, b: Double, min: Double, max: Double)(randBasis: RandBasis) extends Rand[Double] {
  if (min >= max)
    fatal(s"min $min must be less than max $max")

  if (min < 0)
    fatal(s"min $min must be at least 0")

  if(max > 1)
    fatal(s"max $min must be at most 1")

  private val beta = new Beta(a, b)(randBasis)

  // rejection sampling
  override def draw(): Double = {
    var drawn = beta.draw()

    while(drawn < min || drawn > max) {
      drawn = beta.draw()
    }

    drawn
  }
}

File: python/hail/stats/distributions.py
from hail.history import *
from hail.typecheck import *
from hail.utils.java import Env


class BetaDist(HistoryMixin):
    """
    Represents a
    `beta distribution <https://en.wikipedia.org/wiki/Beta_distribution>`__
    with parameters a and b.
    """

    @typecheck_method(a=numeric,
                      b=numeric)
    @record_init
    def __init__(self, a, b):
        self.a = a
        self.b = b

    def _jrep(self):
        return Env.hail().stats.BetaDist.apply(float(self.a), float(self.b))


class UniformDist(HistoryMixin):
    """
    Represents a uniform distribution on the interval [min, max].
    """

    @typecheck_method(min=numeric,
                      max=numeric)
    @record_init
    def __init__(self, min, max):
        if min >= max:
            raise ValueError("min must be less than max")
        self.min = min
        self.max = max

    def _jrep(self):
        return Env.hail().stats.UniformDist.apply(float(self.min), float(self.max))


class TruncatedBetaDist(HistoryMixin):
    """
    Represents a truncated
    `beta distribution <https://en.wikipedia.org/wiki/Beta_distribution>`__
    with parameters a and b and support [min, max]. Draws are made
    via rejection sampling, i.e. returning the first draw from Beta(a,b) that falls in [min, max].
    This procedure may be slow if the probability mass of Beta(a,b) over [min, max] is small.
    """
    @typecheck_method(a=numeric,
                      b=numeric,
                      min=numeric,
                      max=numeric)
    @record_init
    def __init__(self, a, b, min, max):
        if min >= max:
            raise ValueError("min must be less than max")
        elif min < 0:
            raise ValueError("min cannot be less than 0")
        elif max > 1:
            raise ValueError("max cannot be greater than 1")

        self.min = min
        self.max = max
        self.a = a
        self.b = b

    def _jrep(self):
        return Env.hail().stats.TruncatedBetaDist.apply(float(self.a), float(self.b), float(self.min), float(self.max))
